### PR TITLE
[Wav2Vec2] Improve SpecAugment function by converting numpy based fun…

### PR DESCRIPTION
**Pull request by punitvara**
…ction to pytorch based function

Implements #10459

# What does this PR do?

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors which may be interested in your PR.

### CODE REVIEW

File: src/transformers/models/wav2vec2/modeling_wav2vec2.py [hunk 1]
```diff
@@ -102,14 +102,16 @@ def _compute_mask_indices(
         if sz - min_len <= num_mask:
             min_len = sz - num_mask - 1
 
-        mask_idc = np.random.choice(sz - min_len, num_mask, replace=False)
-        mask_idc = np.asarray([mask_idc[j] + offset for j in range(len(mask_idc)) for offset in range(lengths[j])])
-        mask_idcs.append(np.unique(mask_idc[mask_idc < sz]))
+        mask_idc = torch.randperm(sz - min_len)[:num_mask]
+        mask_idc = torch.from_numpy(
+            np.asarray([mask_idc[j] + offset for j in range(len(mask_idc)) for offset in range(lengths[j])])
```
Comment (patrickvonplaten) [inline src/transformers/models/wav2vec2/modeling_wav2vec2.py:107]: Let's try to not use `np` here. If we use `torch.from_numpy(...)` - it's not GPU-friendly

File: src/transformers/models/wav2vec2/modeling_wav2vec2.py [hunk 2]
```diff
@@ -102,14 +102,16 @@ def _compute_mask_indices(
...
+        )
+        mask_idcs.append(torch.unique(mask_idc[mask_idc < sz]))
 
-    min_len = min([len(m) for m in mask_idcs])
+    min_len = torch.min(mask_idcs)
     for i, mask_idc in enumerate(mask_idcs):
```
Comment (patrickvonplaten) [inline src/transformers/models/wav2vec2/modeling_wav2vec2.py:112]: Let's try to get rid of the for-loop and do tensor operations only

**Comment by patrickvonplaten** [comment]
We need to run benchmark tests to see by how much the speed improved both on CPU and GPU

File: src/transformers/models/wav2vec2/modeling_wav2vec2.py (hunk 1 above)
Comment (punitvara) [reply src/transformers/models/wav2vec2/modeling_wav2vec2.py]: mask_idc = torch.tensor([mask_idc[j] + offset for j in range(len(mask_idc)) for offset in range(lengths[j])])
 
 I just did this. Is this right ?

File: src/transformers/models/wav2vec2/modeling_wav2vec2.py (hunk 2 above)
Comment (punitvara) [reply src/transformers/models/wav2vec2/modeling_wav2vec2.py:111]: I am not sure how to do this. Something like following but not getting how should I put it ? Can you help here.

mask[i, mask_idc] = [True, torch.randperm(mask_idc)[:min_len] if torch.tensor(mask_idcs).size() > min_len]

**Comment by punitvara** [comment]
@patrickvonplaten Can you please help with above comments ?

**Comment by patrickvonplaten** [comment]
Hey @punitvara,

At the moment, I sadly don't have the time to handle the big chunk of the PR. It would be great if you could try to:

1) Find a way to benchmark your new function on GPU and show that it yields a speed-up in the forward pass compared to the old function

2) Try out some advanced PyTorch indexing to replace the for loops.

Taking a look at those PRs should help you: https://github.com/huggingface/transformers/pull/9600, https://github.com/huggingface/transformers/pull/9453, https://github.com/huggingface/transformers/pull/6064

**Comment by patrickvonplaten** [comment]
Closing due to inactivity. Sorry @punitvara, I saw a lot of interest from other people to open a PR and this one seems to have stalled. Feel free to re-open it and give it a second shot if you want :-)

**Comment by punitvara** [comment]
I got busy into some other work. I will try to work on different issue. If you get any PR, feel free to merge it
